backend/src/main/java/de/bund/digitalservice/ris/norms/application/service/VerkuendungsImportService.java

```
@@ -47,4 +88,200 @@ public UUID storeNormendokumentationspaket(Query query) throws IOException {
+
+    if (!mediaType.includes(MimeType.valueOf("application/zip"))) {
+      throw new NotAZipFileException("Invalid filetype: " + file.getFilename());
+    }
+  }
+
+  private Set<Dokument> findParseAndValidateDokumente(Map<String, byte[]> files) {
+    if (!files.containsKey(RECHTSETZUNGSDOKUMENT_FILENAME)) {
+      throw new MissingRechtsetzungsdokumentException();
+    }
+
+    NormManifestationEli normManifestationEli = null;
+
+    Map<String, Dokument> dokumente = new HashMap<>();
+    Queue<String> dokumenteToProcess = new LinkedList<>();
+    dokumenteToProcess.add(RECHTSETZUNGSDOKUMENT_FILENAME);
+
+    while (!dokumenteToProcess.isEmpty()) {
+      String dokumentName = dokumenteToProcess.poll();
+      log.debug("Processing Dokument {}", dokumentName);
+      var dokument = parseAndValidateDokument(dokumentName, files.get(dokumentName));
+      dokumente.put(dokumentName, dokument);
+
+      if (!dokument.getManifestationEli().getSubtype().equals(dokumentName.split("\\.")[0])) {
+        throw new MismatchBetweenFilenameAndSubtypeException(
+          dokumentName,
+          dokument.getManifestationEli()
+        );
+      }
+
+      if (normManifestationEli == null) {
+        normManifestationEli = dokument.getManifestationEli().asNormEli();
+      }
+
+      if (!dokument.getManifestationEli().asNormEli().equals(normManifestationEli)) {
+        throw new InconsistentEliException(
+          dokumentName,
+          normManifestationEli,
+          dokument.getManifestationEli().asNormEli()
+        );
+      }
+
+      log.debug(
+        "Parsed & validated Dokument {} now looking for referenced dokuments in it",
+        dokument.getManifestationEli()
+      );
+
+      dokumenteToProcess.addAll(
+        dokument
+          .getReferencedDokumenteNames()
+          .stream()
+          // check if we already know about the referenced dokument
+          .filter(Predicate.not(dokumente::containsKey))
```

Is this something we should raise an exception about? A document that was already processed, meaning it was referenced already and it is referenced again? is it "fachlich" possible?

---

ldml_de allows it technically but I don't know if it is something that is also "fachlich" possible. Also as this is "just" for collecting all Dokumente I decided agains raising an exception.